// @syncfusion/ej2-build/src/generators/themes.js
'use strict';

var fs = global.fs = global.fs || require('fs');
var common = global.config = global.config || require('../utils/common.js');
var config = common.config();

// Generate theme files
function generateThemes(themes) {
    var comp = config.styleDependency;
    if (comp === 'none') {
        return;
    }
    for (var i = 0; i < themes.length; i++) {
        var content = '', root = '', component;
        var baseContent = '@import \'ej2-base/styles/' + themes[i] + '.scss\';\n';
        if (!comp.length) {
            if (common.currentPackage === 'ej2-base') {
                content = '@import \'definition/' + themes[i] + '.scss\';\n';
            }
            else {
                content = baseContent;
            }
            createThemeFile(themes[i], content);
        }
        else {
            content = comp.toString() === 'ej2' ? '' : baseContent;
            for (var j = 0; j < comp.length; j++) {
                if (typeof comp[j] === 'string') {
                    content = content + getImportStatement(comp[j], themes[i]);
                    if (j === comp.length - 1) {
                        createThemeFile(themes[i], content);
                    }
                }
                else {
                    content = baseContent;
                    component = Object.keys(comp[j])[0];
                    root = root + '@import \'' + component + '/' + themes[i] + '.scss\';\n';
                    content = content + getComponentStyles(comp[j][component], themes[i]);
                    createThemeFile(themes[i], content, component);
                }
            }
            createThemeFile(themes[i], root, component, true);
        }
    }
}
exports.generateThemes = generateThemes;

// Create theme files based on folder structure
function createThemeFile(themeName, content, component, isRoot) {
    if (content.length) {
        var control = (component && !isRoot) ? component + '/' : '';
        if (common.currentRepo !== 'ej2-samples' && !(new RegExp(config.thirdPartyWords.join('|')).test(common.currentRepo))) {
            content = content + getThemeStyles(themeName, control);
            content = !isRoot ? content + '@import \'all.scss\';\n' : content;
        }
        writeFile('./styles/' + control + themeName + '.scss', content);
    }
}
exports.createThemeFile = createThemeFile;

// Get inner level folder styles
function getComponentStyles(dependent, theme) {
    var content = '';
    for (var i = 0; i < dependent.length; i++) {
        content = content + getImportStatement(dependent[i], theme);
    }
    return content;
}
exports.getComponentStyles = getComponentStyles;

// Get import statement from component and theme name
function getImportStatement(comp, theme) {
    var statement = '';
    var styles = (comp === 'ej2' || comp.indexOf('ej2/') !== -1) ? '' : 'styles/';
    if (comp.indexOf('../') !== -1) {
        statement = '@import \'' + comp + '/' + theme + '.scss\';\n';
    }
    else if (comp.indexOf('/') !== -1) {
        var splitted = comp.split('/');
        statement = '@import \'' + splitted[0] + '/' + styles + splitted[1] + '/' + theme + '.scss\';\n';
    }
    else {
        statement = '@import \'' + comp + '/' + styles + theme + '.scss\';\n';
    }
    return statement;
}
exports.getImportStatement = getImportStatement;

// Get additional styles for component's theme
function getThemeStyles(themeName, control) {
    var themeStyles = '';
    var themeFile = themeName + '-definition.scss';
    if (fs.existsSync('./styles/' + control + '_definition.scss')) {
        themeStyles = '@import \'definition.scss\';\n';
    }
    if (fs.existsSync('./styles/' + control + '_' + themeFile)) {
        themeStyles = themeStyles + '@import \'' + themeFile + '\';\n';
    }
    return themeStyles;
}


function writeFile(path, content) {
    var shelljs = global.shelljs = global.shelljs || require('shelljs');
    var arPath = path.split('/');
    arPath.pop();
    shelljs.mkdir('-p', arPath.join('/'));
    fs.writeFileSync(path, content, 'utf8');
}

exports.getThemeStyles = getThemeStyles;
