// @syncfusion/ej2-popups/src/popup/popup.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
define(["require", "exports", "@syncfusion/ej2-base", "@syncfusion/ej2-base", "@syncfusion/ej2-base", "../common/position", "@syncfusion/ej2-base", "@syncfusion/ej2-base", "@syncfusion/ej2-base", "../common/collision"], function (require, exports, ej2_base_1, ej2_base_2, ej2_base_3, position_1, ej2_base_4, ej2_base_5, ej2_base_6, collision_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var PositionData = (function (_super) {
        __extends(PositionData, _super);
        function PositionData() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        __decorate([
            ej2_base_4.Property('left')
        ], PositionData.prototype, "X", void 0);
        __decorate([
            ej2_base_4.Property('top')
        ], PositionData.prototype, "Y", void 0);
        return PositionData;
    }(ej2_base_1.ChildProperty));
    exports.PositionData = PositionData;
    var CLASSNAMES = {
        ROOT: 'e-popup',
        RTL: 'e-rtl',
        OPEN: 'e-popup-open',
        CLOSE: 'e-popup-close'
    };
    var Popup = (function (_super) {
        __extends(Popup, _super);
        function Popup(element, options) {
            var _this = _super.call(this, options, element) || this;
            _this.fixedParent = false;
            return _this;
        }
        Popup.prototype.onPropertyChanged = function (newProp, oldProp) {
            for (var _i = 0, _a = Object.keys(newProp); _i < _a.length; _i++) {
                var prop = _a[_i];
                switch (prop) {
                    case 'width':
                        ej2_base_1.setStyleAttribute(this.element, { 'width': ej2_base_2.formatUnit(newProp.width) });
                        break;
                    case 'height':
                        ej2_base_1.setStyleAttribute(this.element, { 'height': ej2_base_2.formatUnit(newProp.height) });
                        break;
                    case 'zIndex':
                        ej2_base_1.setStyleAttribute(this.element, { 'zIndex': newProp.zIndex });
                        break;
                    case 'enableRtl':
                        this.setEnableRtl();
                        break;
                    case 'position':
                    case 'relateTo':
                        this.refreshPosition();
                        break;
                    case 'offsetX':
                        var x = newProp.offsetX - oldProp.offsetX;
                        this.element.style.left = (parseInt(this.element.style.left, 10) + (x)).toString() + 'px';
                        break;
                    case 'offsetY':
                        var y = newProp.offsetY - oldProp.offsetY;
                        this.element.style.top = (parseInt(this.element.style.top, 10) + (y)).toString() + 'px';
                        break;
                    case 'content':
                        this.setContent();
                        break;
                    case 'actionOnScroll':
                        if (newProp.actionOnScroll !== 'none') {
                            this.wireScrollEvents();
                        }
                        else {
                            this.unwireScrollEvents();
                        }
                        break;
                }
            }
        };
        Popup.prototype.getModuleName = function () {
            return 'popup';
        };
        Popup.prototype.getPersistData = function () {
            return this.addOnPersist([]);
        };
        Popup.prototype.destroy = function () {
            this.element.classList.remove(CLASSNAMES.ROOT, CLASSNAMES.RTL);
            this.unwireEvents();
            _super.prototype.destroy.call(this);
        };
        Popup.prototype.render = function () {
            this.element.classList.add(CLASSNAMES.ROOT);
            var styles = {};
            if (this.zIndex !== 1000) {
                styles.zIndex = this.zIndex;
            }
            if (this.width !== 'auto') {
                styles.width = ej2_base_2.formatUnit(this.width);
            }
            if (this.height !== 'auto') {
                styles.height = ej2_base_2.formatUnit(this.height);
            }
            ej2_base_1.setStyleAttribute(this.element, styles);
            this.setEnableRtl();
            this.setContent();
            this.wireEvents();
        };
        Popup.prototype.wireEvents = function () {
            if (ej2_base_3.Browser.isDevice) {
                ej2_base_6.EventHandler.add(window, 'orientationchange', this.orientationOnChange, this);
            }
            if (this.actionOnScroll !== 'none') {
                this.wireScrollEvents();
            }
        };
        Popup.prototype.wireScrollEvents = function () {
            if (this.getRelateToElement()) {
                for (var _i = 0, _a = this.getScrollableParent(this.getRelateToElement()); _i < _a.length; _i++) {
                    var parent_1 = _a[_i];
                    ej2_base_6.EventHandler.add(parent_1, 'scroll', this.scrollRefresh, this);
                }
            }
        };
        Popup.prototype.unwireEvents = function () {
            if (ej2_base_3.Browser.isDevice) {
                ej2_base_6.EventHandler.remove(window, 'orientationchange', this.orientationOnChange);
            }
            if (this.actionOnScroll !== 'none') {
                this.unwireScrollEvents();
            }
        };
        Popup.prototype.unwireScrollEvents = function () {
            if (this.getRelateToElement()) {
                for (var _i = 0, _a = this.getScrollableParent(this.getRelateToElement()); _i < _a.length; _i++) {
                    var parent_2 = _a[_i];
                    ej2_base_6.EventHandler.remove(parent_2, 'scroll', this.scrollRefresh);
                }
            }
        };
        Popup.prototype.getRelateToElement = function () {
            var relateToElement = this.relateTo === '' ? document.body : this.relateTo;
            this.setProperties({ relateTo: relateToElement }, true);
            return ((typeof this.relateTo) === 'string') ?
                document.querySelector(this.relateTo) : this.relateTo;
        };
        Popup.prototype.scrollRefresh = function (e) {
            if (this.actionOnScroll === 'reposition') {
                if (!(this.element.offsetParent === e.target ||
                    (this.element.offsetParent && this.element.offsetParent.tagName === 'BODY' &&
                        e.target.parentElement == null))) {
                    this.refreshPosition();
                }
            }
            else if (this.actionOnScroll === 'hide') {
                this.hide();
            }
            if (this.actionOnScroll !== 'none') {
                if (this.getRelateToElement()) {
                    var targetVisible = this.isElementOnViewport(this.getRelateToElement(), e.target);
                    if (!targetVisible && !this.targetInvisibleStatus) {
                        this.trigger('targetExitViewport');
                        this.targetInvisibleStatus = true;
                    }
                    else if (targetVisible) {
                        this.targetInvisibleStatus = false;
                    }
                }
            }
        };
        Popup.prototype.isElementOnViewport = function (relateToElement, scrollElement) {
            var scrollParents = this.getScrollableParent(relateToElement);
            for (var parent_3 = 0; parent_3 < scrollParents.length; parent_3++) {
                if (this.isElementVisible(relateToElement, scrollParents[parent_3])) {
                    continue;
                }
                else {
                    return false;
                }
            }
            return true;
        };
        Popup.prototype.isElementVisible = function (relateToElement, scrollElement) {
            var rect = relateToElement.getBoundingClientRect();
            if (!rect.height || !rect.width) {
                return false;
            }
            if (scrollElement.getBoundingClientRect) {
                var parent_4 = scrollElement.getBoundingClientRect();
                return !(rect.bottom < parent_4.top) &&
                    (!(rect.bottom > parent_4.bottom) &&
                        (!(rect.right > parent_4.right) &&
                            !(rect.left < parent_4.left)));
            }
            else {
                var win = window;
                var windowView = {
                    top: win.scrollY,
                    left: win.scrollX,
                    right: win.scrollX + win.outerWidth,
                    bottom: win.scrollY + win.outerHeight
                };
                var off = position_1.calculatePosition(relateToElement);
                var ele = {
                    top: off.top,
                    left: off.left,
                    right: off.left + rect.width,
                    bottom: off.top + rect.height
                };
                var elementView = {
                    top: windowView.bottom - ele.top,
                    left: windowView.right - ele.left,
                    bottom: ele.bottom - windowView.top,
                    right: ele.right - windowView.left
                };
                return elementView.top > 0
                    && elementView.left > 0
                    && elementView.right > 0
                    && elementView.bottom > 0;
            }
        };
        Popup.prototype.preRender = function () {
        };
        Popup.prototype.setEnableRtl = function () {
            this.reposition();
            this.enableRtl ? this.element.classList.add(CLASSNAMES.RTL) : this.element.classList.remove(CLASSNAMES.RTL);
        };
        Popup.prototype.setContent = function () {
            if (!ej2_base_2.isNullOrUndefined(this.content)) {
                this.element.innerHTML = '';
                if (typeof (this.content) === 'string') {
                    this.element.textContent = this.content;
                }
                else {
                    this.element.appendChild(this.content);
                }
            }
        };
        Popup.prototype.orientationOnChange = function () {
            var _this = this;
            setTimeout(function () {
                _this.refreshPosition();
            }, 200);
        };
        Popup.prototype.refreshPosition = function (target) {
            if (!ej2_base_2.isNullOrUndefined(target)) {
                this.checkFixedParent(target);
            }
            this.reposition();
            this.checkCollision();
        };
        Popup.prototype.reposition = function () {
            var pos;
            var relateToElement = this.getRelateToElement();
            if (typeof this.position.X === 'number' && typeof this.position.Y === 'number') {
                pos = { left: this.position.X, top: this.position.Y };
            }
            else if (relateToElement) {
                var display = this.element.style.display;
                this.element.style.display = 'block';
                pos = this.getAnchorPosition(relateToElement, this.element, this.position, this.offsetX, this.offsetY);
                this.element.style.display = display;
            }
            else {
                pos = { left: 0, top: 0 };
            }
            this.element.style.left = pos.left + 'px';
            this.element.style.top = pos.top + 'px';
        };
        Popup.prototype.getAnchorPosition = function (anchorEle, ele, position, offsetX, offsetY) {
            var eleRect = ele.getBoundingClientRect();
            var anchorRect = anchorEle.getBoundingClientRect();
            var anchor = anchorEle;
            var anchorPos = { left: 0, top: 0 };
            if (ele.offsetParent && ele.offsetParent.tagName === 'BODY' && anchorEle.tagName === 'BODY') {
                anchorPos = position_1.calculatePosition(anchorEle);
            }
            else {
                anchorPos = position_1.calculateRelativeBasedPosition(anchor, ele);
            }
            switch (position.X) {
                default:
                case 'left':
                    break;
                case 'center':
                    if (this.targetType === 'container') {
                        anchorPos.left += (anchorRect.width / 2 - eleRect.width / 2);
                    }
                    else {
                        anchorPos.left += (anchorRect.width / 2);
                    }
                    break;
                case 'right':
                    if (this.targetType === 'container') {
                        anchorPos.left += (anchorRect.width - eleRect.width);
                    }
                    else {
                        anchorPos.left += (anchorRect.width);
                    }
                    break;
            }
            switch (position.Y) {
                default:
                case 'top':
                    break;
                case 'center':
                    if (this.targetType === 'container') {
                        anchorPos.top += (anchorRect.height / 2 - eleRect.height / 2);
                    }
                    else {
                        anchorPos.top += (anchorRect.height / 2);
                    }
                    break;
                case 'bottom':
                    if (this.targetType === 'container') {
                        anchorPos.top += (anchorRect.height - eleRect.height);
                    }
                    else {
                        anchorPos.top += (anchorRect.height);
                    }
                    break;
            }
            anchorPos.left += offsetX;
            anchorPos.top += offsetY;
            return anchorPos;
        };
        Popup.prototype.callFlip = function (param) {
            var relateToElement = this.getRelateToElement();
            collision_1.flip(this.element, relateToElement, this.offsetX, this.offsetY, this.position.X, this.position.Y, this.viewPortElement, param, this.fixedParent);
        };
        Popup.prototype.callFit = function (param) {
            if (collision_1.isCollide(this.element, this.viewPortElement).length !== 0) {
                var data = collision_1.fit(this.element, this.viewPortElement, param);
                this.element.style.left = data.left + 'px';
                this.element.style.top = data.top + 'px';
            }
        };
        Popup.prototype.checkCollision = function () {
            var horz = this.collision.X;
            var vert = this.collision.Y;
            if (horz === 'none' && vert === 'none') {
                return;
            }
            if (horz === 'flip' && vert === 'flip') {
                this.callFlip({ X: true, Y: true });
            }
            else if (horz === 'fit' && vert === 'fit') {
                this.callFit({ X: true, Y: true });
            }
            else {
                if (horz === 'fit') {
                    this.callFit({ X: true, Y: false });
                }
                else if (vert === 'fit') {
                    this.callFit({ X: false, Y: true });
                }
                if (horz === 'flip') {
                    this.callFlip({ X: true, Y: false });
                }
                else if (vert === 'flip') {
                    this.callFlip({ Y: true, X: false });
                }
            }
        };
        Popup.prototype.show = function (animationOptions) {
            var _this = this;
            if (this.zIndex === 1000) {
                this.zIndex = getZindexPartial(this.element);
                ej2_base_1.setStyleAttribute(this.element, { 'zIndex': this.zIndex });
            }
            animationOptions = (!ej2_base_2.isNullOrUndefined(animationOptions) && typeof animationOptions === 'object') ?
                animationOptions : this.showAnimation;
            if (this.collision.X !== 'none' || this.collision.Y !== 'none') {
                ej2_base_1.removeClass([this.element], CLASSNAMES.CLOSE);
                ej2_base_1.addClass([this.element], CLASSNAMES.OPEN);
                this.checkCollision();
                ej2_base_1.removeClass([this.element], CLASSNAMES.OPEN);
                ej2_base_1.addClass([this.element], CLASSNAMES.CLOSE);
            }
            if (!ej2_base_2.isNullOrUndefined(animationOptions)) {
                animationOptions.begin = function () {
                    if (!_this.isDestroyed) {
                        ej2_base_1.removeClass([_this.element], CLASSNAMES.CLOSE);
                        ej2_base_1.addClass([_this.element], CLASSNAMES.OPEN);
                    }
                };
                animationOptions.end = function () {
                    if (!_this.isDestroyed) {
                        _this.trigger('open');
                    }
                };
                new ej2_base_4.Animation(animationOptions).animate(this.element);
            }
            else {
                ej2_base_1.removeClass([this.element], CLASSNAMES.CLOSE);
                ej2_base_1.addClass([this.element], CLASSNAMES.OPEN);
                this.trigger('open');
            }
        };
        Popup.prototype.hide = function (animationOptions) {
            var _this = this;
            animationOptions = (!ej2_base_2.isNullOrUndefined(animationOptions) && typeof animationOptions === 'object') ?
                animationOptions : this.hideAnimation;
            if (!ej2_base_2.isNullOrUndefined(animationOptions)) {
                animationOptions.end = function () {
                    if (!_this.isDestroyed) {
                        ej2_base_1.removeClass([_this.element], CLASSNAMES.OPEN);
                        ej2_base_1.addClass([_this.element], CLASSNAMES.CLOSE);
                        _this.trigger('close');
                    }
                };
                new ej2_base_4.Animation(animationOptions).animate(this.element);
            }
            else {
                ej2_base_1.removeClass([this.element], CLASSNAMES.OPEN);
                ej2_base_1.addClass([this.element], CLASSNAMES.CLOSE);
                this.trigger('close');
            }
        };
        Popup.prototype.getScrollableParent = function (element) {
            this.checkFixedParent(element);
            return getScrollableParent(element, this.fixedParent);
        };
        Popup.prototype.checkFixedParent = function (element) {
            var parent = element.parentElement;
            while (parent && parent.tagName !== 'HTML') {
                var parentStyle = getComputedStyle(parent);
                if (parentStyle.position === 'fixed' && this.element.offsetParent && this.element.offsetParent.tagName === 'BODY') {
                    this.element.style.position = 'fixed';
                    this.fixedParent = true;
                }
                parent = parent.parentElement;
                if (ej2_base_2.isNullOrUndefined(this.element.offsetParent) && parentStyle.position === 'fixed'
                    && this.element.style.position === 'fixed') {
                    this.fixedParent = true;
                }
            }
        };
        __decorate([
            ej2_base_4.Property('auto')
        ], Popup.prototype, "height", void 0);
        __decorate([
            ej2_base_4.Property('auto')
        ], Popup.prototype, "width", void 0);
        __decorate([
            ej2_base_4.Property(null)
        ], Popup.prototype, "content", void 0);
        __decorate([
            ej2_base_4.Property('container')
        ], Popup.prototype, "targetType", void 0);
        __decorate([
            ej2_base_4.Property(null)
        ], Popup.prototype, "viewPortElement", void 0);
        __decorate([
            ej2_base_4.Property({ X: 'none', Y: 'none' })
        ], Popup.prototype, "collision", void 0);
        __decorate([
            ej2_base_4.Property('')
        ], Popup.prototype, "relateTo", void 0);
        __decorate([
            ej2_base_1.Complex({}, PositionData)
        ], Popup.prototype, "position", void 0);
        __decorate([
            ej2_base_4.Property(0)
        ], Popup.prototype, "offsetX", void 0);
        __decorate([
            ej2_base_4.Property(0)
        ], Popup.prototype, "offsetY", void 0);
        __decorate([
            ej2_base_4.Property(1000)
        ], Popup.prototype, "zIndex", void 0);
        __decorate([
            ej2_base_4.Property(false)
        ], Popup.prototype, "enableRtl", void 0);
        __decorate([
            ej2_base_4.Property('reposition')
        ], Popup.prototype, "actionOnScroll", void 0);
        __decorate([
            ej2_base_4.Property(null)
        ], Popup.prototype, "showAnimation", void 0);
        __decorate([
            ej2_base_4.Property(null)
        ], Popup.prototype, "hideAnimation", void 0);
        __decorate([
            ej2_base_4.Event()
        ], Popup.prototype, "open", void 0);
        __decorate([
            ej2_base_4.Event()
        ], Popup.prototype, "close", void 0);
        __decorate([
            ej2_base_4.Event()
        ], Popup.prototype, "targetExitViewport", void 0);
        Popup = __decorate([
            ej2_base_5.NotifyPropertyChanges
        ], Popup);
        return Popup;
    }(ej2_base_4.Component));
    exports.Popup = Popup;
    function getScrollableParent(element, fixedParent) {
        var eleStyle = getComputedStyle(element);
        var scrollParents = [];
        var overflowRegex = /(auto|scroll)/;
        var parent = element.parentElement;
        while (parent && parent.tagName !== 'HTML') {
            var parentStyle = getComputedStyle(parent);
            if (!(eleStyle.position === 'absolute' && parentStyle.position === 'static')
                && overflowRegex.test(parentStyle.overflow + parentStyle.overflowY + parentStyle.overflowX)) {
                scrollParents.push(parent);
            }
            parent = parent.parentElement;
        }
        if (!fixedParent) {
            scrollParents.push(document);
        }
        return scrollParents;
    }
    exports.getScrollableParent = getScrollableParent;
    function getZindexPartial(element) {
        var parent = element.parentElement;
        var parentZindex = [];
        while (parent) {
            if (parent.tagName !== 'BODY') {
                var index = document.defaultView.getComputedStyle(parent, null).getPropertyValue('z-index');
                var position = document.defaultView.getComputedStyle(parent, null).getPropertyValue('position');
                if (index !== 'auto' && position !== 'static') {
                    parentZindex.push(index);
                }
                parent = parent.parentElement;
            }
            else {
                break;
            }
        }
        var childrenZindex = [];
        for (var i = 0; i < document.body.children.length; i++) {
            if (!element.isEqualNode(document.body.children[i])) {
                var index = document.defaultView.getComputedStyle(document.body.children[i], null).getPropertyValue('z-index');
                var position = document.defaultView.getComputedStyle(document.body.children[i], null).getPropertyValue('position');
                if (index !== 'auto' && position !== 'static') {
                    childrenZindex.push(index);
                }
            }
        }
        childrenZindex.push('999');
        var finalValue = parentZindex.concat(childrenZindex);
        return Math.max.apply(Math, finalValue) + 1;
    }
    exports.getZindexPartial = getZindexPartial;
});
